src/helpers.cpp

```diff
@@ -94,27 +94,43 @@ namespace ada::helpers {
     return pos > input.size() ? std::string_view() : input.substr(pos);
   }
 
-  ada_really_inline size_t get_host_delimiter_location(const ada::url& url, std::string_view& view, bool& inside_brackets) noexcept {
-    size_t location = url.is_special() ? view.find_first_of(":[/?\\") : view.find_first_of(":[/?");
-
-    // Next while loop is almost never taken!
-    while((location != std::string_view::npos) && (view[location] == '[')) {
-      location = view.find(']',location);
-      if(location == std::string_view::npos) {
-        inside_brackets = true;
-        /**
-         * TODO: Ok. So if we arrive here then view has an unclosed [,
-         * Is the URL valid???
-         */
-      } else {
-        location = url.is_special() ? view.find_first_of(":[/?\\#", location) : view.find_first_of(":[/?#", location);
+  ada_really_inline std::pair<size_t,bool> get_host_delimiter_location(const bool is_special, std::string_view& view) noexcept {
+    const size_t view_size = view.size();
+    size_t location = 0;
+    bool found_colon = false;
+    // It is marginally better to have two loops.
+    if(is_special) {
+      for (;location < view_size; ++location) {
+        if (view[location] == '[') {
+          location = view.find(']', location);
+          if (location == std::string_view::npos) {
+            break;
+          }
+        } else if (view[location] == ':' || view[location] == '/' || view[location] == '?' || view[location] == '\\') {
```

Comment:
The only difference between those if statements are: is_special has `view[location] == '\\'` and the other doesn't. Can we merge these functions into a single without compromising performance?

Reply:
There might be different ways to skip this cat, but we don't want to be branching unnecessarily on is_special with each character. It is cheap, but it does add instructions and unnecessary (predictable) branching.

I will see whether this can be improved.